src/rust/engine/src/externs/options.rs

```diff
@@ -177,19 +175,50 @@ type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
 // We could get rid of this tuple type by representing the final value and its rank as
 // a singleton derivation (in the case where we don't otherwise need the full derivation).
 // But that would allocate two unnecessary Vecs for every option.
-type OptionValue<T> = (Option<T>, isize, Option<OptionValueDerivation<T>>);
+type OptionValue<'py, T> = (Option<T>, isize, Option<OptionValueDerivation<'py, T>>);
+
+fn source_to_details(source: &Source) -> Option<&str> {
+    match source {
+        Source::Default => None,
+        Source::Config { ordinal: _, path } => Some(path),
+        Source::Env => Some("env var"),
+        Source::Flag => Some("command-line flag"),
+    }
+}
+
+fn to_details<'py>(py: Python<'py>, sources: Vec<&'py Source>) -> Option<Bound<'py, PyString>> {
+    if sources.is_empty() {
+        return None;
+    }
+    if sources.len() == 1 {
+        return source_to_details(sources.first().unwrap()).map(|s| PyString::intern_bound(py, s));
+    }
+    #[allow(unstable_name_collisions)]
+    // intersperse is provided by itertools::Itertools, but is also in the Rust nightly
+    // as an experimental feature of standard Iterator. If/when that becomes standard we
+    // can use it, but for now we must squelch the name collision.
+    Some(PyString::intern_bound(
```

Comment:
We do end up allocating a new Rust String in this case, but it's an uncommon case so that seems fine (and the Python-side interning will still work).